### PR TITLE
Add initial REST API server for VAST

### PR DESCRIPTION
**Pull request by lava**
### :memo: Reviewer Checklist

Review this pull request by ensuring the following items:

- [ ] All user-facing changes have changelog entries
- [ ] User-facing changes are reflected on [vast.io](https://github.com/tenzir/vast/tree/master/web)

### CODE REVIEW

**Comment by dominiklohmann** [comment]
[sc-34483]

File: plugins/rest/src/plugin.cpp
Comment (lava) [inline plugins/rest/src/plugin.cpp:61]: Make this a `component_plugin` as well, so that the server state can be implemented in the component to have shared access and can be persisted in the server.

File: plugins/rest/CMakeLists.txt
Comment (lava) [inline plugins/rest/CMakeLists.txt:26]: @dominiklohmann I'm not sure why this line is required

File: plugins/rest/www/status.html
Comment (lava) [inline plugins/rest/www/status.html:43]: Remove rainbow effect and add a boring status page instead

File: libvast/native-plugins/rest_api/status.cpp
Comment (lava) [inline libvast/native-plugins/rest_api/status.cpp:65]: This needs to be fixed so we ideally get a new handler in the node that directly returns a string, but alternatively we can just use this pattern:

```
    caf::message_handler{[&](const std::string& str) {
      // use string
    }}(e.context());
```

File: plugins/rest/src/server_command.cpp
Comment (lava) [inline plugins/rest/src/server_command.cpp:115]: @dominiklohmann Here we still need to cast value to the type expected to the function.

File: libvast/native-plugins/rest_api/export.cpp
Comment (mavam) [inline libvast/native-plugins/rest_api/export.cpp:275]: Why do we need these IDs? I couldn't infer the use case.

File: Dockerfile
Comment (dominiklohmann) [inline Dockerfile:28]: These need to be added to the instructions for building from source, and need to be installed in CI (where this plugin isn't being built yet).

File: plugins/rest/src/server_command.cpp
Comment (dominiklohmann) [inline plugins/rest/src/server_command.cpp:260]: This will only work when running VAST from the build tree. We should install the www directory somewhere into DATAROOTDIR, I think.

File: plugins/rest/src/server_command.cpp
Comment (dominiklohmann) [inline plugins/rest/src/server_command.cpp:112]: Do we need to implement this before merging, or is this for a follow-up story?

File: plugins/rest/src/generate_token_command.cpp
Comment (dominiklohmann) [inline plugins/rest/src/generate_token_command.cpp:25]: That name is also used for the top-level entry in `vast status`, so the question becomes whether users find `vast status rest` or `vast status authenticator` easier.

File: plugins/rest/src/authenticator.cpp
Comment (dominiklohmann) [inline plugins/rest/src/authenticator.cpp:18]: Is there a reason not to use the C++ standard library tools here?

File: plugins/rest/include/rest/restinio_response.hpp
Comment (dominiklohmann) [inline plugins/rest/include/rest/restinio_response.hpp:39]: Probably makes sense given that it's rather likely for the object to be in an unusable state after calling abort.

File: libvast/test/rest_api.cpp
Comment (dominiklohmann) [inline libvast/test/rest_api.cpp:99]: That can probably be done with `from_yaml`, which parses YAML strings into a `data`.

File: libvast/include/vast/type.hpp
Comment (dominiklohmann) [inline libvast/include/vast/type.hpp:1393]: Adding this for concrete types looks rather dangerous to me. CAF's type inspection only works well for semi-regular types, and record type isn't semi-regular, so this will never be usable unless you wrap it into an optional. That means that we have no way to test this inspect handler.

Could you change the code that is using this to use a `type` rather than a `optional<record_type>`?

File: libvast/include/vast/plugin.hpp
Comment (dominiklohmann) [inline libvast/include/vast/plugin.hpp:366]: How much effort do you think it would be to refactor this to not require plugin developers to write their own actor?

File: libvast/include/vast/plugin.hpp
Comment (dominiklohmann) [inline libvast/include/vast/plugin.hpp:345]: Naming question: it's the REST endpoint plugin, the REST handler actor, but a set of API endpoint. Can you add a comment on how they relate, ideally somewhere in the plugin API?

File: libvast/native-plugins/rest_api/export.cpp
Comment (dominiklohmann) [inline libvast/native-plugins/rest_api/export.cpp:29]: If you comment the handlers similar to how we do it in actors.hpp you won't need this.

File: plugins/rest/include/rest/restinio_response.hpp
Comment (dominiklohmann) [inline plugins/rest/include/rest/restinio_response.hpp:21]: What for?

File: plugins/rest/include/rest/restinio_response.hpp
Comment (dominiklohmann) [inline plugins/rest/include/rest/restinio_response.hpp:27]: Since users are supposed to work with this I think it makes sense to comment this well, especially the override functions.

File: plugins/rest/include/rest/restinio_response.hpp
Comment (dominiklohmann) [inline plugins/rest/include/rest/restinio_response.hpp:47]: Do we understand why that is?

File: plugins/rest/src/authenticator.cpp
Comment (dominiklohmann) [inline plugins/rest/src/authenticator.cpp:86]: Open TODO.

File: plugins/rest/src/authenticator.cpp
Comment (dominiklohmann) [inline plugins/rest/src/authenticator.cpp:91]: Open TODO.

File: plugins/rest/src/plugin.cpp
Comment (dominiklohmann) [inline plugins/rest/src/plugin.cpp:37]: Open FIXME

File: plugins/rest/src/plugin.cpp
Comment (dominiklohmann) [inline plugins/rest/src/plugin.cpp:44]: In terms of naming, do you think it makes sense to call this `web` instead, since it's more powerful than just a REST API?

File: plugins/rest/src/plugin.cpp
Comment (dominiklohmann) [inline plugins/rest/src/plugin.cpp:75]: These all need documentation, ideally both in a README.md and on vast.io.

File: plugins/rest/CMakeLists.txt
Comment (dominiklohmann) [inline plugins/rest/CMakeLists.txt:5]: This plugin is yet lacking a CHANGELOG.md file that matches this version.

File: plugins/rest/CMakeLists.txt
Comment (dominiklohmann) [inline plugins/rest/CMakeLists.txt:1]: I think upstream is at 3.24 now for the policy opt-in version.

File: plugins/rest/include/rest/configuration.hpp
Comment (dominiklohmann) [inline plugins/rest/include/rest/configuration.hpp:42]: Unaddressed FIXME.

File: libvast/native-plugins/rest_api/export.cpp
Comment (dominiklohmann) [inline libvast/native-plugins/rest_api/export.cpp:1]: Would it maybe make more sense to bundle these native plugins as part of the REST plugin itself? I think that can be implemented in our CMake with some shuffling around.

The URL should probably also match plugin name, i.e., `rest_endpoints` rather than `rest_api`.

**Review by dominiklohmann — COMMENTED** [review]
I looked at the code more in depth and left a bunch of comments on things that I think should be looked at.

There are no major issues with the code. It'd be nice if the REST endpoints didn't have to implement a whole actor interface, but I don't know how feasible that actually is.

The biggest remaining piece of work is probably documentation, both for the usage of the REST plugin and the plugin API of the REST endpoints.

I also gave this a quick spin locally, and it looks like it's working fine. Once we addressed the remaining comments I think we can quickly approve and merge this. I'm also really interesting in seeing how this integrates with #2574.

File: plugins/rest/src/server_command.cpp
Comment (dominiklohmann) [reply plugins/rest/src/server_command.cpp]: This is fixed for at least basic types, but there's still an outstanding TODO for lists. Did you want to address that still, or is that waiting for the story that is about implementing "parse this string as a data with this type information" more generically?

File: plugins/rest/CMakeLists.txt
Comment (dominiklohmann) [reply plugins/rest/CMakeLists.txt:26]: I cannot reproduce this locally, but looking at the snippet where we link against it there are some other cleanups we can do:

https://github.com/tenzir/vast/blob/e42a9652e50833c5ba53bbd34d6ff12892a1cfa4/libvast/CMakeLists.txt#L432-L443

We actually require {fmt} >= 7 afaict, so maybe it's got something to do with that?

File: plugins/rest/src/server_command.cpp
Comment (lava) [reply plugins/rest/src/server_command.cpp:112]: It's a nice-to-have, but all parameters are currently correctly documented as query-parameters so it's certainly not a blocker.

File: libvast/test/rest_api.cpp
Comment (lava) [reply libvast/test/rest_api.cpp:99]: Sadly, the comment is incorrect and it's JSON data instead of YAML.

File: plugins/rest/src/authenticator.cpp
Comment (dominiklohmann) [reply plugins/rest/src/authenticator.cpp:18]: Fair point. I think we already link against OpenSSL via CAF, so maybe we can use that for it?

File: libvast/include/vast/type.hpp
Comment (lava) [reply libvast/include/vast/type.hpp:1393]: Hm, `optional<record_type>` is the semantically correct type so if we use `type` then we'd have to downcast and assert that it's a record on the call site. Could we instead allow empty `record_type` to make it semi-regular? Or alternatively, we could just create an `inspect(Inspector&, caf::optional<record_type>&)` overload to make sure it only works if the record type is correctly wrapped.

Comment (dominiklohmann) [reply libvast/include/vast/type.hpp:1393]: We cannot make concrete types semi-regular, that would break quite a few assumptions that we can currently make about record types.

I'm fine with adding an inspect overload like this:

```cpp
template <class Inspector, concrete_type Type>
auto inspect(Inspector& f, std::optional<Type>& x)
    -> typename Inspector::result_type {
  return f(...);
}
```

(Actually not sure if that works generically with concrete type because of the usual ADL wonkiness, but if it doesn't then maybe just for `record_type` as a friend.)

File: libvast/include/vast/plugin.hpp
Comment (lava) [reply libvast/include/vast/plugin.hpp:366]: From the plugin server PoV it would be easy to have the plugin provide a function like 

```
class rest_endpoint_plugin {
public:
  using handler = std::function<void(http_request&&)>;

  virtual handler get_handler() const = 0;
};
```

and to call that when handling the request.

The problem is that on the plugin developer side it's pretty difficult to provide any useful functionality unless you can interact with the actor system.

File: plugins/rest/include/rest/restinio_response.hpp
Comment (lava) [reply plugins/rest/include/rest/restinio_response.hpp:47]: Yes, that comment is technically outdated; restinio has some optimizations where doesn't make a copy when given a string_view so that it becomes cheap to return static string literals, but that breaks when passing a regular string as `string_view`.

File: plugins/rest/src/plugin.cpp
Comment (lava) [reply plugins/rest/src/plugin.cpp:44]: Yes, makes sense to me

File: plugins/rest/include/rest/restinio_response.hpp
Comment (lava) [reply plugins/rest/include/rest/restinio_response.hpp:27]: The user-facing comments are in `vast::http_response`, the `restinio_response` is an implementation detail of the rest plugin.

File: libvast/native-plugins/rest_api/export.cpp
Comment (lava) [reply libvast/native-plugins/rest_api/export.cpp:1]: Either way makes sense to me; I mainly split them out of the rest plugin so I don't accidentally rely on internals of the REST plugin in the implementation, which would lead to future problems for closed-source endpoints.

File: plugins/rest/src/authenticator.cpp
Comment (lava) [reply plugins/rest/src/authenticator.cpp:91]: I think this one is fine leave for now until we also add some ways to list and display the generated tokens.

File: plugins/rest/src/server_command.cpp
Comment (lava) [reply plugins/rest/src/server_command.cpp]: I'll probably create a separate story for supporting lists, so we can implement it when we actually want to add the first endpoint with a list parameter.

File: plugins/rest/src/authenticator.cpp
Comment (lava) [reply plugins/rest/src/authenticator.cpp:18]: Good point, recent version of OpenSSL use `getentropy()` on Linux anyways to implement `RANDOM_bytes()`

File: libvast/include/vast/plugin.hpp
Comment (dominiklohmann) [reply libvast/include/vast/plugin.hpp:366]: I see. Maybe this is something that we can abstract away in the future by making it easier to interact with the actor system. Feel free to resolve this for now.

File: plugins/rest/src/plugin.cpp
Comment (mavam) [reply plugins/rest/src/plugin.cpp:75]: Documentation on vast.io is a separate story. The README will just point there.

(IMO this can be resolved.)

File: plugins/rest/src/generate_token_command.cpp
Comment (lava) [reply plugins/rest/src/generate_token_command.cpp:25]: Ideally we'd have a separate component called `rest` that displays the actually interesting state (ie. listen address and mode), but I'm not sure that's implementable with the current plugin framework.

File: libvast/include/vast/system/actors.hpp
Comment (lava) [reply libvast/include/vast/system/actors.hpp:423]: I agree it makes sense to create strong typedef's for all the various id's that we're sending as messages, but currently we use uint64/uint32 for everything else as well so that's probably a separate refactoring.

File: plugins/rest/src/server_command.cpp
Comment (lava) [reply plugins/rest/src/server_command.cpp]: I've removed the html part for now, we can add it back in a later commit.

File: libvast/native-plugins/rest_api/export.cpp
Comment (lava) [reply libvast/native-plugins/rest_api/export.cpp]: I did the renaming, I didn't do the reshuffling for now since at least then we have all native plugins in one place.

File: libvast/native-plugins/rest_api/export.cpp
Comment (lava) [reply libvast/native-plugins/rest_api/export.cpp]: I know, but I still don't like the idea of adding more work to fool a tool into behaving as if it was turned off, instead of just turning it off :D

File: plugins/web/CHANGELOG.md
Comment (netantho) [inline plugins/web/CHANGELOG.md:3]: Typo: **web** plugin?

Comment (lava) [reply plugins/web/CHANGELOG.md]: Of course, thanks :D

**Review by dominiklohmann — APPROVED** [review]
We peer-reviewed this. Approving this as-is modulo the remaining comments. We have a few things left to do in follow-ups.

**Pull request merged**
